RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemoverPhase2.cc

```diff
@@ -0,0 +1,358 @@
+  }
+}
+
+
+SeedClusterRemoverPhase2::~SeedClusterRemoverPhase2()
+{
+}
+
+void SeedClusterRemoverPhase2::mergeOld(ClusterRemovalInfo::Indices &refs,
+                                            const ClusterRemovalInfo::Indices &oldRefs) 
+{
+        for (size_t i = 0, n = refs.size(); i < n; ++i) {
+            refs[i] = oldRefs[refs[i]];
+       }
+}
+
+ 
+template<typename T> 
+auto_ptr<edmNew::DetSetVector<T> >
+SeedClusterRemoverPhase2::cleanup(const edmNew::DetSetVector<T> &oldClusters, const std::vector<uint8_t> &isGood, 
+			     reco::ClusterRemovalInfo::Indices &refs, const reco::ClusterRemovalInfo::Indices *oldRefs){
+    typedef typename edmNew::DetSetVector<T>             DSV;
+    typedef typename edmNew::DetSetVector<T>::FastFiller DSF;
+    typedef typename edmNew::DetSet<T>                   DS;
+    auto_ptr<DSV> output(new DSV());
+    output->reserve(oldClusters.size(), oldClusters.dataSize());
+
+    unsigned int countOld=0;
+    unsigned int countNew=0;
+
+    // cluster removal loop
+    const T * firstOffset = & oldClusters.data().front();
+    for (typename DSV::const_iterator itdet = oldClusters.begin(), enddet = oldClusters.end(); itdet != enddet; ++itdet) {
+        DS oldDS = *itdet;
+
+        if (oldDS.empty()) continue; // skip empty detsets
+
+        uint32_t id = oldDS.detId();
+        DSF outds(*output, id);
+
+        for (typename DS::const_iterator it = oldDS.begin(), ed = oldDS.end(); it != ed; ++it) {
+            uint32_t index = ((&*it) - firstOffset);
+	    countOld++;
+            if (isGood[index]) { 
+                outds.push_back(*it);
+		countNew++;
+                refs.push_back(index); 
+                LogDebug("SeedClusterRemoverPhase2") << "SeedClusterRemoverPhase2::cleanup " << typeid(T).name() << " reference " << index << " to " << (refs.size() - 1);
+            }
+
+	}
+        if (outds.empty()) outds.abort(); // not write in an empty DSV
+    }
+
+    if (oldRefs != 0) mergeOld(refs, *oldRefs);
+    return output;
+}
+
+
+
+
+void SeedClusterRemoverPhase2::process(const TrackingRecHit *hit, float chi2, const TrackerGeometry* tg) {
+
+  DetId detid = hit->geographicalId(); 
+  uint32_t subdet = detid.subdetId();
+
+  assert(subdet > 0);
+
+  const type_info &hitType = typeid(*hit);
+  if (hitType == typeid(SiPixelRecHit)) {
+
+    if(!doPixel_) return;
+
+    const SiPixelRecHit *pixelHit = static_cast<const SiPixelRecHit *>(hit);
+    SiPixelRecHit::ClusterRef cluster = pixelHit->cluster();
+    LogDebug("SeedClusterRemoverPhase2") << "Plain Pixel RecHit in det " << detid.rawId();
+
+    if (cluster.id() != pixelSourceProdID) throw cms::Exception("Inconsistent Data") << 
+            "SeedClusterRemoverPhase2: pixel cluster ref from Product ID = " << cluster.id() << 
+            " does not match with source cluster collection (ID = " << pixelSourceProdID << ")\n.";
+
+    assert(cluster.id() == pixelSourceProdID);
```

Comment:
As I said, I'm against this practice in general. Nevertheless, this area of code is not my responsibility. I was asked to get a full reco workflow going ASAP, and this is what I was given to work with in the short term. The people responsible for this code have responded already.